modules/packages/Sort.chpl

```diff
@@ -1155,6 +1155,252 @@ module ShellSort {
   }
 }
 
+
+pragma "no doc"
+module SampleSortHelp {
+  param maxLogBuckets = 8; // not counting equality buckets.
+  param classifyUnrollFactor = 7;
+  const equalBucketThreshold = 5;
+  const maxInline = 16;
+  const oversampleFactor = 0.2;
+
+  // compute logarithm base 2 rounded down
+  proc log2int(n: int) {
+    if n <= 0 then
+      return 0;
+    return log2(n);
+  }
+
+  proc computeLogBucketSize(n: int) {
+    const LogBuckets = 8;
+    const BaseCaseSize = 16;
+    const BaseCaseMultiplier = 16;
+    const SingleLevelThreshold = maxInline * LogBuckets;
+    const TwoLevelThreshold = SingleLevelThreshold * LogBuckets;
+    var ret = 0;
+    if n <= SingleLevelThreshold {
+      ret = log2int(n / maxInline / 2);
+    } else if n <= TwoLevelThreshold {
+      ret = (log2int(n / maxInline / 2)+1)/2;
+    } else {
+      ret = maxLogBuckets;
+    }
+
+    ret = max(1, ret); // make sure it's at least 1
+    ret = min(maxLogBuckets, ret); // make sure it's at most maxLogBuckets.
+
+    return ret;
+  }
+  proc chooseSampleStep(n:int, logNumBuckets:int) {
+    var oversample = min(1.0, oversampleFactor * log2int(n));
+    var step = max(1, oversample:int);
+    return step;
+  }
+
+
+
+  record SampleBucketizer {
+    type eltType;
+
+    // filled from 1 to num_buckets_
+    var storage: c_array(eltType, 1<<maxLogBuckets);
+    // filled from 0 to num_buckets, last one is duplicated
+    var sortedStorage: c_array(eltType, 1<<maxLogBuckets);
+
+    var logBuckets: int;
+    var numBuckets: int;
+    var equalBuckets: bool;
+
+    proc writeThis(ch) {
+      ch <~> "SampleBucketizer(";
+      ch <~> "\n logBuckets=" <~> logBuckets;
+      ch <~> "\n numBuckets=" <~> numBuckets;
+      ch <~> "\n equalBuckets=" <~> equalBuckets;
+      ch <~> "\n storage=";
+      for i in 0..numBuckets {
+        ch <~> (try! " %xt".format(storage[i]));
+      }
+      ch <~> "\n sortedStorage=";
+      for i in 0..numBuckets {
+        ch <~> (try! " %xt".format(sortedStorage[i]));
+      }
+      ch <~> ")\n";
+    }
+
+    proc getNumBuckets() {
+      return numBuckets * (1 + equalBuckets:int);
+    }
```

Comment:
If we had a Chapel linter, it would probably complain to you about inconsistent spacing (newlines) between function definitions.